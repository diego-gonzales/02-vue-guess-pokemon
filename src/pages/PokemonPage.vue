<template>
  <h1 v-if="!correctPokemon">
    Please wait a moment...
  </h1>
  <div v-else>
    <h1>Who is that pokemon?</h1>
    <PokemonPicture :pokemon-id="correctPokemon.id"
                    :show-pokemon="showPokemon"
    />
    <PokemonOptions :pokemons="pokemonsArray"
                    @selection-pokemon="checkAnswer"
    />

    <template v-if="showAnswer">
      <h2>
        {{ message }}
      </h2>
      <button @click="newGame">
        New game
      </button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from '../components/PokemonPicture';
import PokemonOptions from '../components/PokemonOptions';
import getPokemonOptions from '@/helpers/getPokemonOptions';

export default {
  name: 'PokemonPage',
  components: {
    PokemonPicture,
    PokemonOptions
  },
  data() {
    return {
      pokemonsArray: [],
      correctPokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async getPokemons() {
      this.pokemonsArray = await getPokemonOptions();
      const randomNumber = Math.floor(Math.random() * 4);

      this.correctPokemon = this.pokemonsArray[randomNumber];
    },
    checkAnswer(pokemonIdEvent) {
      this.showPokemon = true;
      this.showAnswer = true;
      (this.correctPokemon.id === pokemonIdEvent)
          ? this.message = `Correct!`
          : this.message = `Incorrect!. Its ${this.correctPokemon.name}`;
    },
    newGame() {
      this.pokemonsArray = []
      this.correctPokemon = null
      this.showPokemon = false
      this.showAnswer = false
      this.getPokemons();
    }
  },
  mounted() {
    // Podr√≠amos decir que el mounted() es equivalente al ngOnInit() en Angular
    this.getPokemons();
  }
}
</script>