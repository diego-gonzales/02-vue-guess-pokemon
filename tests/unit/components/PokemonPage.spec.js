import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { mockArrayPokemons } from '../mocks/pokemons.mock'


describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    // Este snapshot no servirá de mucho ya que solo muestra el loading
    test('should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('should call getPokemons() when my app has been mounted', () => {
        /* Para esta parte ver video 108 para más detalle. El spy debe definirse antes del shallowMount()
        ya que cuando hago el spy entonces ya es muy tarde para espiar el metodo getPokemons(), debido a que
        el componente ya esta totalmente montado y el getPokemns() solo estará disponible en el momento en que se está
        montando la app */
        const getPokemonsSpy = jest.spyOn(PokemonPage.methods, 'getPokemons')
        shallowMount(PokemonPage)

        expect(getPokemonsSpy).toHaveBeenCalled()
    })

    /* Esta sería otra forma de tener un snapshot completo de mi componente (EXPLICADO MEJOR EN EL VIDEO 109,
    vale la pena verlo para conocer la diferencia entre SHALLOWMOUNT() y MOUNT()) */
    test('should match with snapshot when load all pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: mockArrayPokemons,
                    correctPokemon: mockArrayPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    optionColor: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot();
    })

    test('should display the PokemonPicture and PokemonOptions components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: mockArrayPokemons,
                    correctPokemon: mockArrayPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    optionColor: ''
                }
            }
        })

        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        const pokemonOptions = wrapper.find('pokemon-options-stub')

        expect(pokemonPicture.exists()).toBeTruthy()
        expect(pokemonOptions.exists()).toBeTruthy()
        expect(pokemonPicture.attributes('pokemonid')).toBe('1') // que sea iguala uno porque ese es el id del pokemon en la posicion 0, ver la data enviada arriba
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy()
    })

    test('should trigger checkAnswer() method', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: mockArrayPokemons,
                    correctPokemon: mockArrayPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: '',
                    optionColor: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(mockArrayPokemons[0].id) // colocamos await ya que ese metodo hará modificaciones al DOM

        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBe(true);
        expect(wrapper.find('h2').text()).toBe('Correct!')

        await wrapper.vm.checkAnswer(7)
    
        expect(wrapper.vm.message).toBe(`Incorrect!. Its ${mockArrayPokemons[0].name}`)
    })
    
})
