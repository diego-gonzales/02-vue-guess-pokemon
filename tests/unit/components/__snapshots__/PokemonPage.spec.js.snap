// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component should match with snapshot 1`] = `
<h1>
  <pokemon-loader-stub></pokemon-loader-stub>
</h1>
`;

exports[`PokemonPage Component should match with snapshot when load all pokemons 1`] = `
<div>
  <h1>Who is that pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" optioncolor=""></pokemon-options-stub>
  <!--v-if-->
</div>
`;
